src/EventStore.Core/TransactionLog/Chunks/TFChunkWriter.cs

```diff
@@ -40,24 +43,50 @@ public class TFChunkWriter : ITransactionFileWriter {
 			// DO NOTHING
 		}
 
+		public void Begin() {
+			if (_inTransaction)
+				throw new Exception("Attempt to begin a new transaction while there is an ongoing transaction.");
+
+			_inTransaction = true;
+		}
+
 		public bool Write(ILogRecord record, out long newPos) {
 			var result = _currentChunk.TryAppend(record);
 			if (result.Success)
-				_writerCheckpoint.Write(result.NewPosition + _currentChunk.ChunkHeader.ChunkStartPosition);
+				_nextRecordPosition = result.NewPosition + _currentChunk.ChunkHeader.ChunkStartPosition;
 			else
-				CompleteChunk(); // complete updates checkpoint internally
-			newPos = _writerCheckpoint.ReadNonFlushed();
+				CompleteChunk();
+
+			if (!_inTransaction)
+				CommitInternal();
+
+			newPos = _nextRecordPosition;
 			return result.Success;
 		}
 
+		public void Commit() {
+			if (!_inTransaction)
+				throw new Exception("Attempt to commit while there is no ongoing transaction.");
+
+			_inTransaction = false;
+			CommitInternal();
+		}
+
+		private void CommitInternal(bool flush = false) {
```

Comment:
personally i think this would be better if the parameter was non-optional, and be explicit about the intention on every call (theres only four of them anyway)